projects/documentation/content/migrations/2023-6-11.md

```diff
@@ -0,0 +1,97 @@
+
+```ts
+html`
+    <element-containing-an-overlay
+        @sp-closed=${(event: OverlayOpenCloseDetail) => {
+            if (event.detail.interaction === 'auto') {
+                // Do something when the event was dispatched for an overlay with type "auto".
+            }
+        }}
+        @sp-opened=${(event: OverlayOpenCloseDetail) => {
+            if (event.detail.interaction === 'modal') {
+                // Do something when the event was dispatched for an overlay with type "modal".
+            }
+        }}
+    ></element-containing-an-overlay>
+`;
+```
+
+### Prepare for descendant overlays to exist in the same DOM tree
+
+Previously, the reparenting of overlay content prevented `sp-opened` and `sp-closed` events from bubbling as would normally be the default. This means you will have the opportunity to hear and respond to these events ALL in a centralized location, but your listeners may hear more of these events than you expect and/or had previously.
+
+✅ _DO_ be sure to gate your listeners if your experience stacks multiple overlays (see submenus) within each other.
+
+```ts
+function handleSpOpened(event: OverlayOpenCloseDetail) {
+    // Return if the `sp-opened` event was not dispatched from the element to which this listener is attached.
+    if (event.target !== event.currentTarget) return;
+}
+```
+
+<script type="module">
+    import '@spectrum-web-components/dialog/sp-dialog.js';
+    import '@spectrum-web-components/overlay/overlay-trigger.js';
+    import '@spectrum-web-components/popover/sp-popover.js';
+    import '@spectrum-web-components/tooltip/sp-tooltip.js';
+</script>
+
+## Explanation
+
+The new version of the Overlay API no longer relies on portalling (moving content to the end of the `<body>` element) to defeat CSS clipping and stacking. While this approach was good at overcoming these realities, the reparenting of the overlaid content required to apply this technique had high-performance costs for constantly reorienting elements to a new parent and broke encapsulation, making it difficult to take full control of the delivery of that content as far as styling. In exchange, the new API will leverage `<dialog>` elements and the `showModal()` method for modal overlays and the `popover` attribute along with the `showPopover()` method they add to elements to which they are applied. Both of these APIs lift content onto the [top layer](https://developer.mozilla.org/en-US/docs/Glossary/Top_layer) of the browser which provides a full guarantee against CSS clipping and stacking interrupting the content addressed to this layer. Be aware that content on this layer is managed as a strict stack, so the last added will always be "on top" regardless of any additional CSS you may apply. [A discussion](https://github.com/adobe/spectrum-web-components/discussions/2764#discussioncomment-5327797) around additional features that could address this reality is ongoing, please jump in with your thoughts if you have them.
```

Comment:
> The new version of the Overlay API no longer relies on portalling (moving content to the end of the `<body>` element) to defeat CSS clipping and stacking. While this approach was good at overcoming those realities, the reparenting of the overlaid content required to apply this technique had high performance costs, because it was constantly reorienting elements to a new parent and broke encapsulation. This made it difficult to take full control of styling and delivering overlaid content. In exchange, the new API will leverage `<dialog>` elements and the `showModal()` method for modal overlays, and the `popover` attribute along with the `showPopover()` method added to elements to which they are applied. Both of these APIs lift content onto the [top layer](https://developer.mozilla.org/en-US/docs/Glossary/Top_layer) of the browser, which provides a full guarantee against CSS clipping and stacking interrupting the content addressed to this layer. Be aware that content on this layer is managed as a strict stack, so the element added last will always be "on top", regardless of any additional CSS you may apply. [A discussion](https://github.com/adobe/spectrum-web-components/discussions/2764#discussioncomment-5327797) around additional features that could address this reality is ongoing; please jump in with your thoughts if you have them.

Reply:
I think these changes make the content here clearer to understand.